src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,678 @@
+            set => SetValue(key, value);
+        }
+
+        public TValue this[int index]
+        {
+            get => GetAt(index);
+            set => SetAt(index, value);
+        }
+
+        public OrderedDictionary()
+            : this(0, null)
+        {
+        }
+
+        public OrderedDictionary(int capacity)
+            : this(capacity, null)
+        {
+        }
+
+        public OrderedDictionary(IEqualityComparer<TKey> comparer)
+            : this(0, comparer)
+        {
+        }
+
+        public OrderedDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+            if (capacity > 0)
+            {
+                Resize(HashHelpers.GetPrime(capacity));
+            }
+
+            if (comparer != EqualityComparer<TKey>.Default)
+            {
+                _comparer = comparer;
+            }
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection)
+            : this(collection, null)
+        {
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey> comparer)
+            : this((collection as ICollection<KeyValuePair<TKey, TValue>>)?.Count ?? 0, comparer)
+        {
+            if (collection == null)
+            {
+                throw new ArgumentNullException(nameof(collection));
+            }
+
+            foreach (KeyValuePair<TKey, TValue> pair in collection)
+            {
+                Add(pair.Key, pair.Value);
+            }
+        }
+
+        public void Add(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.ThrowOnExisting);
+
+        public void Clear()
+        {
+            int count = Count;
+            if (count > 0)
+            {
+                Array.Clear(_buckets, 0, _buckets.Length);
+                Array.Clear(_entries, 0, count);
+                Count = 0;
+            }
+        }
+
+        public bool ContainsKey(TKey key) => IndexOf(key) >= 0;
+
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (_entries.Length >= capacity)
+            {
+                return _entries.Length;
+            }
+            int newSize = HashHelpers.GetPrime(capacity);
+            Resize(newSize);
+            ++_version;
+            return newSize;
+        }
+
+        public TValue GetAt(int index) => GetAt(index, out _);
+
+        public TValue GetAt(int index, out TKey key)
+        {
+            if ((uint)index >= (uint)Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), Strings.ArgumentOutOfRange_Index);
+            }
+
+            Entry entry = _entries[index];
```

Comment:
It doesn't check if the entry has been removed? Is it a part of the contract?

Reply:
When an entry is removed all the entries after it are pushed down so that the entries array is always contiguous.